src/fem/fe_system.cc

```diff
@@ -0,0 +1,416 @@
+    // in base FE, first come the nodal dofs, then cell dofs
+    for (unsigned int i=0; i<n_cell_dofs; i++)
+      cell_basis.push_back(basis_offset+fe->n_object_dofs(0,DOF_SINGLE)+i);
+    
+    fe_index++;
+    comp_offset += fe->n_components();
+    basis_offset += fe->n_dofs();
+  }
+  
+  // make the permutation of dofs: first nodal dofs, then cell dofs
+  for (unsigned int i=0; i<dim+1; i++)
+    dof_basis_.insert(dof_basis_.end(), node_basis[i].begin(), node_basis[i].end());
+  dof_basis_.insert(dof_basis_.end(), cell_basis.begin(), cell_basis.end());
+  
+//   if (this->is_primitive_)
+//   {
+//     double dof_index = 0;
+//     // add index of nonzero component for each dof in FESystem
+//     for (auto fe : fe_)
+//     {
+//       for (int i=0; i<fe->n_dofs(); ++i)
+//         this->component_indices_.push_back(fe_dof_indices_[dof_index++].fe_index);
+//     }
+//   } else {
+    double dof_index = 0;
+    comp_offset = 0;
+    // add footprint of nonzero components for each dof in FESystem
+    for (auto fe : fe_)
+    {
+      for (int i=0; i<fe->n_dofs(); ++i)
+      {
+        std::vector<bool> nonzeros(n_components_, false);
+        for (unsigned int c=0; c<fe->n_components(); c++)\
+          nonzeros[comp_offset+c] = fe->get_nonzero_components(fe_dof_indices_[dof_index++].basis_index)[c];
+        this->nonzero_components_.push_back(nonzeros);
+      }
+      comp_offset += fe->n_components();
+    }
+//   }
+
+  compute_node_matrix();
+}
+
+
+
+template<unsigned int dim, unsigned int spacedim>
+double FESystem<dim,spacedim>::basis_value(const unsigned int i, const arma::vec::fixed<dim> &p) const
+{
+  OLD_ASSERT(i <= number_of_dofs, "Index of basis function is out of range.");
+  return fe_[fe_dof_indices_[dof_basis_[i]].fe_index]->basis_value(fe_dof_indices_[dof_basis_[i]].basis_index, p);
+}
+
+template<unsigned int dim, unsigned int spacedim>
+arma::vec::fixed<dim> FESystem<dim,spacedim>::basis_grad(const unsigned int i, const arma::vec::fixed<dim> &p) const
+{
+  OLD_ASSERT(i <= number_of_dofs, "Index of basis function is out of range.");
+  return fe_[fe_dof_indices_[dof_basis_[i]].fe_index]->basis_grad(fe_dof_indices_[dof_basis_[i]].basis_index, p);
+}
+
+template<unsigned int dim, unsigned int spacedim>
+double FESystem<dim,spacedim>::basis_value_component(const unsigned int i, 
+                                                     const arma::vec::fixed<dim> &p, 
+                                                     const unsigned int comp) const
+{
+  OLD_ASSERT(i <= number_of_dofs, "Index of basis function is out of range.");
+  
+  unsigned int bi = dof_basis_[i];
+  // component index in the base FE
+  int l_comp = comp-fe_dof_indices_[bi].component_offset;
+  OLD_ASSERT(l_comp >= 0 && l_comp < fe_[fe_dof_indices_[bi].fe_index]->n_components(),
+    "Index of component is out of range.");
+
+  if (fe_[fe_dof_indices_[bi].fe_index]->n_components() == 1)
```

Comment:
So this can be simplified if we have only this method and not the previous one.

Reply:
Postponed to #887

Reply:
This seems simple make it now.

Reply:
Done